### ISSUE TITLE
✅ 🐞 Invalid "yes" unit breaks 3D transform with css "--e-transform-rotateX:1yes;"

### ISSUE THREAD

**Issue by Erudition** [elementor/elementor#29339]
### Prerequisites

- [X] I have searched for similar issues in open and closed tickets and cannot find a duplicate.
- [X] I have troubleshooted my issue, and it still exists against the latest stable version of Elementor.

### Description

My 3D transformed elements stopped working, and I've been digging around for days to find out why they work in the Elementor editor but not in the frontend page. Then I discovered the slightest difference between the styles applied to each:

Editor:
```css
elementor-8387 .elementor-element.elementor-element-d8367cd > .elementor-widget-container {
  --e-transform-rotateZ:0deg;
  --e-transform-rotateX:25deg;
  --e-transform-perspective:1000px;
  --e-transform-rotateY:17deg;
}
```
Frontend:
```css
.elementor-8387 .elementor-element.elementor-element-d8367cd > .elementor-widget-container {
  --e-transform-rotateZ:0deg;
  --e-transform-rotateX:1yes;
  --e-transform-rotateY:17deg;
  --e-transform-perspective:20px;
}
```

What the heck is a "yes" unit? At first I thought it was legitimately a new css unit I didn't know about, but once I saw it was degrees in the editor, I looked it up and confirmed that only degrees are valid here.

Was some sort of string replacement going on? Well, I disabled all plugins one by one and the "yes" didn't go away. Clearing cache, refreshing elementor styles, etc.

Finally I started digging into the Elementor codebase. I don't know PHP, but I found something interesting.

Commit https://github.com/elementor/elementor/commit/8b8d3e31258bd04858251113815104e8b30a2ed6 changed a lot, but it was labeled a "Tweak" and none of the PR form was filled out. It moved around the 3D transform controls, turning **this line from up to v3.9.0-beta1**:

https://github.com/elementor/elementor/blob/623508ed172c82565998700b6e7d34c28d170155/includes/widgets/common.php#L744

with emphasis on
```
'transform-rotateX: 1{{UNIT}};  --e-' . $transform_css_modifier . 'transform-perspective: 20px;',
```

Into **this line as of v3.9.0-beta2**:

https://github.com/elementor/elementor/blob/69deef127429ecc2dfeef37103ef334e529ece52/includes/base/element-base.php#L918

with emphasis on
```
'--e-transform-rotateX: 1deg;  --e-transform-perspective: 20px;',
```

The {{UNIT}} syntax is valid because it's used in the neighboring transformX and transformY controls, but those controls have unit selectors and default units specified. Transform3D doesn't...

So UNIT is getting filled with something from somewhere. Could it be here, in the defaults for a SWITCHER control?

https://github.com/elementor/elementor/blob/69deef127429ecc2dfeef37103ef334e529ece52/includes/controls/switcher.php#L75

```php
	protected function get_default_settings() {
		return [
			'label_off' => esc_html__( 'No', 'elementor' ),
			'label_on' => esc_html__( 'Yes', 'elementor' ),
			'return_value' => 'yes',
		];
	}
```

### Steps to reproduce

Turn on 3D Transform on a widget.
Observe that the widget tilts in X and Y as you adjust the angles in the editor.

Publish, and view the page on the frontend. Observe that the transform is not applied.

To make things more confusing, Firefox nor Chrome dev tools show the css property as being overriden, nor as being invalid, yet the computed tab shows `transform` is computed to `none`.

In the inspector, observe the CSS variables for the transform, e.g. :

```css
.elementor-8387 .elementor-element.elementor-element-d8367cd > .elementor-widget-container {
  --e-transform-rotateZ:0deg;
  --e-transform-rotateX:1yes;
  --e-transform-rotateY:17deg;
  --e-transform-perspective:20px;
}

.e-transform .elementor-widget-container {
	transform: perspective(var(--e-transform-perspective,0)) rotate(var(--e-transform-rotateZ,0)) rotateX(var(--e-transform-rotateX,0)) rotateY(var(--e-transform-rotateY,0)) translate(var(--e-transform-translate,0)) translateX(var(--e-transform-translateX,0)) translateY(var(--e-transform-translateY,0)) scaleX(calc(var(--e-transform-flipX, 1) * var(--e-transform-scaleX, var(--e-transform-scale, 1)))) scaleY(calc(var(--e-transform-flipY, 1) * var(--e-transform-scaleY, var(--e-transform-scale, 1)))) skewX(var(--e-transform-skewX,0)) skewY(var(--e-transform-skewY,0));
	transform-origin: var(--e-transform-origin-y) var(--e-transform-origin-x);
}
```

### Expected behavior

The transform works both in the editor and out, as it did in past versions. Invalid CSS units are not generated.

### Isolating the problem

- [ ] This bug happens when only the Elementor (and Elementor Pro) plugins are active.
- [X] This bug happens with the Hello Elementor theme active.
- [X] I can reproduce this bug consistently by following the steps I described above.

### Elementor System Info

<details>
<summary>Click to reveal</summary>

````txt

== Server Environment ==
	Operating System: Linux
	Software: h2o/2.2.6
	MySQL version: MySQL Community Server (GPL) v5.7.12
	PHP Version: 8.1.30
	PHP Memory Limit: 1024M
	PHP Max Input Vars: 5000
	PHP Max Post Size: 518M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: All right
	Elementor Library: Connected

== WordPress Environment ==
	Version: 6.6.2
	Site URL: https://9c6ceb32-c7b3-4df6-9572-1cc6e129636e.static.getshifter.net
	Home URL: https://9c6ceb32-c7b3-4df6-9572-1cc6e129636e.static.getshifter.net
	WP Multisite: No
	Max Upload Size: 512 MB
	Memory limit: 256M
	Max Memory limit: 512M
	Permalink Structure: /%postname%/
	Language: en_US
	Timezone: America/Chicago
	Debug Mode: Inactive

== Theme ==
	Name: Hello Elementor
	Version: 3.1.1
	Author: Elementor Team
	Child Theme: No

== User ==
	Role: administrator
	WP Profile lang: en-US
	User Agent: Mozilla/5.0 (X11; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0

== Active Plugins ==
	Elementor
		Version: 3.25.8
		Author: Elementor.com

	Elementor Pro
		Version: 3.25.3
		Author: Elementor.com

	Plugin Notes Plus
		Version: 1.2.8
		Author: Jamie Bergen

	Pods - Custom Content Types and Fields
		Version: 3.2.8
		Author: Pods Framework Team


== Must-Use Plugins ==
	EVA
		Version: 1.7.1
		Author: DigitalCube


== Elements Usage ==
	
	archive : 3
		archive-posts : 1
		button : 1
		container : 15
		heading : 7
		lae-portfolio : 1
		loop-grid : 2
		search-form : 1
		sitemap : 2
		text-editor : 3
	container : 3
		button : 8
		container : 57
		elementskit-countdown-timer : 1
		heading : 16
		icon-box : 37
		image : 16
		text-editor : 3
	livemesh_grid : 0
		livemesh-grid-item : 10
	livemesh_item : 0
		container : 2
		image : 2
	page : 8
		accordion : 1
		animated-headline : 1
		button : 11
		countdown : 4
		elementskit-page-list : 2
		gallery : 1
		heading : 49
		icon : 2
		icon-box : 1
		icon-list : 4
		image : 6
		image-box : 10
		image-carousel : 2
		lae-portfolio : 2
		nav-menu : 1
		social-icons : 2
		spacer : 8
		table-of-contents : 4
		text-editor : 29
		theme-archive-title : 1
		video : 4
	popup : 2
		button : 2
		container : 3
		heading : 2
		html : 1
		image : 1
		nav-menu : 1
		text-editor : 1
	section : 28
		animated-headline : 3
		button : 17
		countdown : 6
		counter : 11
		divider : 2
		ekit-nav-menu : 1
		elementskit-countdown-timer : 2
		elementskit-icon-box : 10
		gallery : 1
		heading : 45
		icon : 2
		icon-list : 13
		image : 8
		image-box : 4
		lottie : 1
		sitemap : 1
		spacer : 2
		table-of-contents : 1
		text-editor : 21
		theme-site-logo : 1
		toggle : 2
		video : 1
	widget : 0
		elementskit-page-list : 1
		gallery : 1
		icon-box : 1
		table-of-contents : 1
		text-editor : 1
	wp-post : 2
		button : 12
		container : 31
		elementskit-heading : 10
		elementskit-icon-box : 13
		elementskit-page-list : 14
		heading : 12
		icon-box : 4
		image : 6
		image-box : 1
		text-editor : 1
	wp-page : 40
		accordion : 1
		alert : 3
		animated-headline : 4
		button : 114
		call-to-action : 1
		container : 160
		countdown : 1
		counter : 12
		divider : 16
		elementskit-accordion : 1
		elementskit-blog-posts : 1
		elementskit-countdown-timer : 2
		elementskit-image-box : 3
		elementskit-page-list : 7
		elementskit-testimonial : 1
		flip-box : 3
		form : 1
		gallery : 11
		google_maps : 1
		heading : 239
		html : 2
		icon : 26
		icon-box : 70
		icon-list : 7
		image : 82
		image-box : 31
		image-carousel : 3
		lae-posts-carousel : 1
		lae-testimonials : 1
		lae-testimonials-slider : 1
		loop-carousel : 1
		loop-grid : 7
		lottie : 1
		media-carousel : 2
		mega-menu : 1
		social-icons : 2
		spacer : 35
		table-of-contents : 17
		tabs : 1
		template : 1
		testimonial-carousel : 3
		text-editor : 191
		toggle : 4
		video : 12
	single-post : 2
		alert : 2
		container : 9
		counter : 1
		heading : 5
		image : 1
		post-info : 2
		post-navigation : 1
		search-form : 1
		sitemap : 1
		spacer : 1
		text-editor : 2
		theme-post-content : 1
		theme-post-title : 1
	loop-item : 3
		container : 3
		image : 3
		post-info : 2
		text-editor : 2
		theme-post-title : 1
	error-404 : 1
		button : 3
		heading : 1
		html : 1
		image : 1
		search-form : 1
		text-editor : 2


== Settings ==
	
	cpt_support: post, page
	allow_tracking: yes
	unfiltered_files_upload: 1
	form-submissions: 1
	meta_generator_tag: 1


== Features ==
	Custom Fonts: 0
	Custom Icons: 0

== Integrations ==
	


== Elementor Experiments ==
	Improved CSS Loading: Inactive by default
	Inline Font Icons: Inactive by default
	Additional Custom Breakpoints: Active
	Container: Active
	Upgrade Swiper Library: Active
	Nested Elements Performance: Active by default
	Optimized Control Loading: Active
	Optimized Markup: Inactive
	Conditionally load Swiper CSS files: Inactive by default
	Plugin Onboarding: Active by default
	CSS Smooth Scroll: Active by default
	Default to New Theme Builder: Active by default
	Header & Footer: Active
	Elementor Home Screen: Active by default
	Landing Pages: Active by default
	Nested Elements: Active
	Editor Top Bar: Active
	Pages Panel: Inactive by default
	Element Caching: Inactive by default
	Link In Bio: Active by default
	Floating Buttons: Active by default
	Elementor Editor Events: Inactive by default
	Atomic Widgets: Inactive by default
	Launchpad Checklist: Inactive by default
	Menu: Active


== Log ==
	
PHP: showing 6 of 6PHP: 2024-05-24 12:04:06 [notice X 116][../wp-content/plugins/elementor/includes/base/controls-stack.php::229] hexdec(): Passing null to parameter #1 ($hex_string) of type string is deprecated [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2024-06-03 17:46:14 [warning X 1][../wp-content/plugins/elementor-pro/modules/query-control/module.php::768] Attempt to read property "name" on null [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2024-06-10 16:52:31 [warning X 4][../wp-content/plugins/elementor/modules/history/revisions-manager.php::363] Undefined array key "editor_post_id" [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2024-10-03 16:29:47 [error X 1][../wp-content/plugins/elementor/core/files/uploads-manager.php::546] Uncaught TypeError: Elementor\Core\Files\Uploads_Manager::validate_file(): Argument #1 ($file) must be of type array, null given, called in ../wp-content/plugins/elementor/core/files/uploads-manager.php on line 192 and defined in ../wp-content/plugins/elementor/core/files/uploads-manager.php:546
Stack trace:
#0 ../wp-content/plugins/elementor/core/files/uploads-manager.php(192): Elementor\Core\Files\Uploads_Manager->validate_file(NULL)
#1 ../wp-includes/class-wp-hook.php(324): Elementor\Core\Files\Uploads_Manager->handle_elementor_wp_media_upload(NULL)
#2 ../wp-includes/plugin.php(205): WP_Hook->apply_filters(NULL, Array)
#3 ../wp-admin/includes/file.php(833): apply_filters('wp_handle_uploa...', NULL)
#4 ../wp-admin/includes/file.php(1105): _wp_handle_upload(NULL, Array, '2024-10-03 16:2...', 'wp_handle_uploa...')
#5 ../wp-admin/includes/media.php(306): wp_handle_upload(NULL, Array, '2024-10-03 16:2...')
#6 ../wp-admin/includes/ajax-actions.php(2632): media_handle_upload('async-upload', '2852', Array)
#7 ../wp-admin/async-upload.php(33): wp_ajax_upload_attachment()
#8 {main}
  thrown [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2024-10-22 20:51:00 [warning X 13][../wp-content/plugins/elementor/includes/base/controls-stack.php::695] Undefined array key -1 [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2024-10-23 09:34:08 [warning X 12][../wp-content/plugins/elementor-pro/core/app/modules/site-editor/data/endpoints/templates.php::150] Undefined array key "condition_type" [array (
  'trace' => '
#0: ../wp-content/plugins/elementor-pro/core/app/modules/site-editor/data/endpoints/templates.php(150): Elementor\Core\Logger\Manager -> rest_error_handler()
#1: ElementorPro\Core\App\Modules\SiteEditor\Data\Endpoints\Templates -> normalize_template_json_item()
#2: ../wp-content/plugins/elementor-pro/core/app/modules/site-editor/data/endpoints/templates.php(120): class type array_map()
#3: ../wp-content/plugins/elementor-pro/core/app/modules/site-editor/data/endpoints/templates.php(59): ElementorPro\Core\App\Modules\SiteEditor\Data\Endpoints\Templates -> normalize_templates_json()
#4: ../wp-content/plugins/elementor/data/base/endpoint.php(158): ElementorPro\Core\App\Modules\SiteEditor\Data\Endpoints\Templates -> get_items()
',
)]

JS: showing 11 of 11JS: 2024-08-29 20:28:08 [error X 4][../wp-content/plugins/elementor-pro/assets/js/editor.min.js?ver=3.21.1:2:65789] can't convert undefined to object 
JS: 2024-10-21 22:42:22 [error X 82][../wp-content/plugins/elementor/assets/js/editor.min.js?ver=3.21.3:3:683613] elementorFrontend.elements.window.jQuery is not a function 
JS: 2024-10-21 22:42:22 [error X 22][../wp-includes/js/mediaelement/mediaelement-and-player.min.js?ver=4.2.17:12:0]  
JS: 2024-10-21 22:55:06 [error X 1][../wp-content/plugins/elementor/assets/js/editor.min.js?ver=3.21.3:3:821067] T.originalEvent.key is undefined 
JS: 2024-10-23 02:09:42 [error X 19][../wp-content/plugins/ooohboi-steroids-for-elementor/assets/js/ooohboi-steroids.js?ver=2.1.607072022:11:56] Cannot read properties of undefined (reading 'handlers') 
JS: 2024-10-23 02:09:42 [error X 19][../wp-includes/js/jquery/jquery.min.js?ver=3.7.1:2:28760] Cannot read properties of undefined (reading 'tools') 
JS: 2024-10-23 14:34:53 [error X 23][../wp-content/plugins/elementor/assets/js/frontend.min.js?ver=3.24.7:2:7352] elementorModules.frontend is undefined 
JS: 2024-10-23 16:44:08 [error X 11][../wp-content/plugins/elementor/assets/js/frontend-modules.min.js?ver=3.24.7:2:17144] e is undefined 
JS: 2024-10-23 16:44:09 [error X 85][../wp-content/plugins/elementor/assets/js/frontend-modules.min.js?ver=3.24.7:2:17144] undefined has no properties 
JS: 2024-10-23 18:56:50 [error X 1][../wp-content/plugins/elementor/assets/js/editor.min.js?ver=3.24.7:3:154943] Cannot read properties of null (reading 'scrollTo') 
JS: 2024-11-20 06:18:10 [error X 1][../wp-content/plugins/elementor/assets/js/editor.min.js?ver=3.25.8:3:1274404] $ is null 

Log: showing 20 of 252024-10-22 17:08:06 [info] elementor-pro::elementor_pro_updater Started 
2024-10-22 17:08:06 [info] Elementor Pro/Upgrades - _on_each_version Start  
2024-10-22 17:08:06 [info] Elementor Pro/Upgrades - _on_each_version Finished 
2024-10-22 17:08:06 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.21.1',
  'to' => '3.24.4',
)]
2024-10-22 17:08:06 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.21.1',
  'to' => '3.24.4',
)]
2024-11-13 11:58:57 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.24.4',
  'to' => '3.25.3',
)]
2024-11-13 11:58:59 [info] elementor-pro::elementor_pro_updater Started 
2024-11-13 11:58:59 [info] Elementor Pro/Upgrades - _on_each_version Start  
2024-11-13 11:58:59 [info] Elementor Pro/Upgrades - _on_each_version Finished 
2024-11-13 11:58:59 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.24.4',
  'to' => '3.25.3',
)]
2024-11-13 11:59:40 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.24.7',
  'to' => '3.25.6',
)]
2024-11-13 11:59:41 [info] elementor::elementor_updater Started 
2024-11-13 11:59:41 [info] Elementor/Upgrades - _on_each_version Start  
2024-11-13 11:59:44 [info] Elementor/Upgrades - _on_each_version Finished 
2024-11-13 11:59:44 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.24.7',
  'to' => '3.25.6',
)]
2024-11-19 23:55:33 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.25.6',
  'to' => '3.25.8',
)]
2024-11-19 23:55:34 [info] elementor::elementor_updater Started 
2024-11-19 23:55:34 [info] Elementor/Upgrades - _on_each_version Start  
2024-11-19 23:55:37 [info] Elementor/Upgrades - _on_each_version Finished 
2024-11-19 23:55:37 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.25.6',
  'to' => '3.25.8',
)]



== Elementor - Compatibility Tag ==
	
	Elementor Pro: Compatible

== Elementor Pro - Compatibility Tag ==

````

</details>

### Agreement

- [X] I agree that my issue may be closed without action if it doesn't meet all the requirements.

**Comment by jamieburchell** [comment]
@Erudition timely release...

> **3.25.9 – 2024-11-20**
> _Fix: 3D Rotate transform is not functioning as expected when the Optimized Control Loading feature is activated_

Maybe that fixes it.

**Comment by nicholaszein** [comment]
Hi @Erudition!

We have great news! 🙌 

### This issue was solved with the release of Elementor version 3.25.9! 

✅ Feel free to check it out and update your plugin to the new version!

> _Check the changelog for more information:_
> [Changelog of Elementor and Elementor Pro](https://elementor.com/help/changelog-of-elementor/)

Cheers 🥂

Ps: Thank you @jamieburchell for confirming the fix. 🙏
